File: androidApp/src/main/java/com/example/ula_app/android/ui/game/DebugScreen.kt
package com.example.ula_app.android.ui.game

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.text.ClickableText
import androidx.compose.material.Button
import androidx.compose.material.ButtonDefaults
import androidx.compose.material.DropdownMenu
import androidx.compose.material.DropdownMenuItem
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.example.ula_app.android.data.DataSource
import com.example.ula_app.android.ui.viewmodel.DebugViewModel
import com.example.ula_app.android.ui.viewmodel.HomeViewModel

private const val TAG = "DebugScreen"

private const val leftWeight = 0.3f
private const val rightWeight = 1 - leftWeight

@Composable
fun DebugScreen(
        homeViewModel: HomeViewModel = viewModel(),
        debugViewModel: DebugViewModel = viewModel()
) {
    val ageOptions = listOf(
        "",
        DataSource.MonsterAgeOptions.Egg.name,
        DataSource.MonsterAgeOptions.Child.name,
        DataSource.MonsterAgeOptions.Adult.name
    )

    val bodyStatusOptions = listOf(
        "",
        DataSource.MonsterBodyStatusOptions.NA.name,
        DataSource.MonsterBodyStatusOptions.Normal.name,
        DataSource.MonsterBodyStatusOptions.Overweight.name,
        DataSource.MonsterBodyStatusOptions.Fat.name,
        DataSource.MonsterBodyStatusOptions.Fit.name,
    )


    val movieIdsOptions = mutableListOf<String>()
    movieIdsOptions.add("")
    movieIdsOptions += DataSource.eggRule + DataSource.childNormalRule + DataSource.childFatRule +
            DataSource.childOverweightRule + DataSource.adultNormalRule + DataSource.adultFatRule +
            DataSource.adultFitRule + DataSource.adultOverweightRule

    var ageOptionExpanded by remember {
        mutableStateOf(false)
    }

    var bodyStatusOptionExpanded by remember {
        mutableStateOf(false)
    }

    var movieIdOptionExpanded by remember {
        mutableStateOf(false)
    }

    var ageOptionSelectedIndex by remember {
        mutableStateOf(ageOptions[0])
    }
    var bodyStatusOptionSelectedIndex by remember {
        mutableStateOf(bodyStatusOptions[0])
    }
    var movieIdOptionSelectedIndex by remember {
        mutableStateOf(movieIdsOptions[0])
    }

    Column(
        modifier = Modifier
            .fillMaxSize(),
//        horizontalArrangement = Arrangement.Center,
    ) {

        // Specify the title
        Text(
            text = "Debug and Stay ALIVE!",
            fontSize = 30.sp
        )

        Spacer(modifier = Modifier.height(35.dp))

        // Create dropdown boxes
        Column(
                modifier = Modifier
                    .fillMaxWidth(0.8f),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Dropdown(
                selectedIndex = ageOptionSelectedIndex,
                expand = ageOptionExpanded,
                dropdownTitle = "Age",
                dropdownOptions = ageOptions,
                onDropdownClicked = {
                    ageOptionExpanded = it
                },
                onDropdownItemClicked = {
                    ageOptionSelectedIndex = it
                }
            )

            Spacer(modifier = Modifier.height(8.dp))

            Dropdown(
                selectedIndex = bodyStatusOptionSelectedIndex,
                expand = bodyStatusOptionExpanded,
                dropdownTitle = "Body Status",
                dropdownOptions = bodyStatusOptions,
                onDropdownClicked = {
                    bodyStatusOptionExpanded = it
                },
                onDropdownItemClicked = {
                    bodyStatusOptionSelectedIndex = it
                }
            )

            Spacer(modifier = Modifier.height(8.dp))

            Dropdown(
                selectedIndex = movieIdOptionSelectedIndex,
                expand = movieIdOptionExpanded,
                dropdownTitle = "movie Id",
                dropdownOptions = movieIdsOptions,
                onDropdownClicked = {
                    movieIdOptionExpanded = it
                },
                onDropdownItemClicked = {
                    movieIdOptionSelectedIndex = it
                }
            )

            Button(
                onClick = {
                    if(debugViewModel.isOnlyChangeId(ageOptionSelectedIndex, bodyStatusOptionSelectedIndex, movieIdOptionSelectedIndex)){
                        // change the movie with id only
                        homeViewModel.setId(movieIdOptionSelectedIndex)
                    } else {
                        // change the age and body status
                        homeViewModel.setAgeAndBodyStatus(ageOptionSelectedIndex, bodyStatusOptionSelectedIndex)
                    }
                },
                colors = ButtonDefaults.outlinedButtonColors(contentColor = Color.Black),
                enabled = debugViewModel.isEnabledButton(ageOptionSelectedIndex, bodyStatusOptionSelectedIndex, movieIdOptionSelectedIndex)

            ) {
                Text(text = "Change the movie! ")
            }
        }
    }
}

@Composable
fun Dropdown(
    selectedIndex: String,
    expand: Boolean,
    dropdownTitle: String,
    dropdownOptions: List<String>,
    onDropdownClicked: (Boolean) -> Unit,
    onDropdownItemClicked: (String) -> Unit,
    enable: Boolean = true
) {
    Row(
        modifier = Modifier
                .fillMaxWidth()
                .height(24.dp)
    ) {
        Column(
            modifier = Modifier
                    .weight(leftWeight)
                    .fillMaxHeight(),
            verticalArrangement = Arrangement.Center,
            horizontalAlignment = Alignment.Start
        ) {
            Text(text = "${dropdownTitle}")
        }
        Column(
            modifier = Modifier
                    .weight(rightWeight)
                    .fillMaxHeight(),
            verticalArrangement = Arrangement.Center,
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            ClickableText(
                //text = AnnotatedString(dropdownOptions[selectedIndex]), // get the options text using the index
                text = AnnotatedString(selectedIndex),
                style = TextStyle(
                    textAlign = TextAlign.Center,
                    color = Color.White
                ),
                onClick = {

                    if(enable) {
                        onDropdownClicked(true)
                    }
                },
                modifier = Modifier
                        .fillMaxWidth()
                        .fillMaxHeight()
                        .background(Color.DarkGray)
            )
            DropdownMenu(
                expanded = expand,
                onDismissRequest = { onDropdownClicked(false) },
                modifier = Modifier
                        .fillMaxWidth(0.56f)
                        .background(Color.White)
            ) {

                dropdownOptions.forEachIndexed { index, s ->
                    DropdownMenuItem(
                        onClick = {
                            onDropdownItemClicked(s)
                            onDropdownClicked(false)
                        },
                        modifier = Modifier
                            .fillMaxWidth()

                    ) {
                        Text(text = "$s")
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/game/ui/DebugScreen.swift
//
//  DebugScreen.swift
//  iosApp
//
//  Created by Hao Wu on 8/19/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared
import SwiftUI

struct DebugScreen: View {
    
    @ObservedObject var debugViewModel: IOSDebugViewModel
    @ObservedObject var homeViewModel: IOSHomeViewModel
    
    init() {
        self.debugViewModel = IOSDebugViewModel()
        self.homeViewModel = IOSHomeViewModel()
    }
    
    let ageOptions = [
        "",
        DataSource.MonsterAgeOptions.egg.name,
        DataSource.MonsterAgeOptions.child.name,
        DataSource.MonsterAgeOptions.adult.name
    ]
    
    let bodyStatusOptions = [
        "",
        DataSource.MonsterBodyStatusOptions.na.name,
        DataSource.MonsterBodyStatusOptions.normal.name,
        DataSource.MonsterBodyStatusOptions.overweight.name,
        DataSource.MonsterBodyStatusOptions.fat.name,
        DataSource.MonsterBodyStatusOptions.fit.name,
    ]
    
    
    @State private var ageOptionSelected = ""
    @State private var bodyStatusOptionSelected = ""
    @State private var movieIdOptionSelected = ""
    
    var body: some View {
        
        VStack {
            Text("Debug and Stay ALIVE!")
                .font(.AppBody)
            
            Spacer()
            
            Picker("Age: ", selection: $ageOptionSelected) {
                ForEach(ageOptions, id: \.self) {
                    Text($0).font(.AppBody)
                }
            }.pickerStyle(.menu).font(.AppBody)
            
            Picker("Body Status: ", selection: $bodyStatusOptionSelected) {
                ForEach(bodyStatusOptions, id: \.self) {
                    Text($0).font(.AppBody)
                }
            }.pickerStyle(.menu).font(.AppBody)
            
            Picker("Movie ID: ", selection: $movieIdOptionSelected) {
                ForEach(debugViewModel.getMovieIdOptions(), id: \.self) {
                    Text($0).font(.AppBody)
                }
            }.pickerStyle(.menu).font(.AppBody)
            
            Spacer()
            
            Button("Change the movie! ", action: {

                if debugViewModel.isOnlyChangeId(ageSelection: ageOptionSelected, bodyStatusSelection: bodyStatusOptionSelected, movieSelection: movieIdOptionSelected) {
                    
                    homeViewModel.setId(id: movieIdOptionSelected)
                    print("age: ", ageOptionSelected)
                    print("body status: ", bodyStatusOptionSelected)
                    print("movie id: ", movieIdOptionSelected)
                    
                } else {
                    homeViewModel.setAgeAndBodyStatus(age: ageOptionSelected, bodyStatus: bodyStatusOptionSelected)
                    print("age: ", ageOptionSelected)
                    print("body status: ", bodyStatusOptionSelected)
                    print("movie id: ", movieIdOptionSelected)
                    
                }
            })
            .disabled(!debugViewModel.isEnableButton(ageSelection: ageOptionSelected, bodyStatusSelection: bodyStatusOptionSelected, movieSelection: movieIdOptionSelected))
            .font(.AppBody)
            
            Spacer()
            
        }
        
    }
    
}


//struct DebugScreen_Previews: PreviewProvider {
//    static var previews: some View {
//        DebugScreen()
//    }
//}
